core/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -199,17 +204,27 @@ public NodeEnvironment(Settings settings, Environment environment) throws IOExce
             IOException lastException = null;
             int maxLocalStorageNodes = MAX_LOCAL_STORAGE_NODES_SETTING.get(settings);
             for (int possibleLockId = 0; possibleLockId < maxLocalStorageNodes; possibleLockId++) {
-                for (int dirIndex = 0; dirIndex < environment.dataFiles().length; dirIndex++) {
-                    Path dataDirWithClusterName = environment.dataWithClusterFiles()[dirIndex];
-                    Path dataDir = environment.dataFiles()[dirIndex];
+                for (int dirIndex = 0; dirIndex < environment.dataFiles().length + extra; dirIndex++) {
+                    final Path dataDir;
+                    if (dirIndex < environment.dataFiles().length) {
```

Comment:
So I looked into this when I first wrote the code here and I think there's way to much going on to extract a method here *as part of this pull request*. I've wanted to refactor this code in the past. Yet, there are too many things being mutated, locks that are released under failure conditions only to be retired on the next go around, etc. that a *straightforward* refactoring is not going to happen. To a certain extent, the code being introduced here is a hack and it's going to be removed again. I really do not want to increase the scope of review here to refactor a loop that is already deserving of being refactored, and the code here is going to look the same when we yank this change back out at some point in the future.